File: JINSEOP/프로그래머스_WEEK14/가장 긴 팰린드롬.cpp
#include <iostream>
#include <string>
using namespace std;

int DP[2501][2501] = {0, };
void init_DP(string s){
    for(int i = 0; i < s.size(); i++)
            DP[i][i] = 1;
}

void find_2byte_palindrome(string s){
    for(int i = 0; i < s.size() - 1; i++)
        if(s[i] == s[i+1])
            DP[i][i+1] = 2;
}

void find_palindrome(string s){
    for(int len = 3; len <= s.size(); len++){
        for(int left = 0; left <= (s.size() - len); left++){
            int right = left + len - 1;
            
            if(s[left] == s[right] && DP[left + 1][right - 1] != 0)
                DP[left][right] = len;
        }
    }
}

int solution(string s)
{
    int answer=1;
   
    init_DP(s);
    find_2byte_palindrome(s);
    find_palindrome(s);
    
    for(int i = 0; i < s.size(); i++){
        for(int j = 0; j < s.size(); j++){
            if(DP[i][j] == 0 || i == j) continue;
            answer = max(DP[i][j], answer);
        }
    }
    
    return answer;
}

File: DONGUK/프로그래머스_WEEK_14/가장 긴 팰린드롬.py
def isPalindrome(input_str):
    N = len(input_str)
    
    if N == 1:
        return True
    
    left, right = '', ''
    if N % 2 == 1:
        left = input_str[0:N//2]
        right = input_str[(N//2)+1:]
    else:
        left = input_str[0:N//2]
        right = input_str[N//2:]

    right = right[::-1]
    flag = True
    for i in range(len(left)):
        if left[i] != right[i]:
            flag = False
            break
            
    return flag
    

def solution(s):
    for cut in range(len(s), 0, -1):
        for start in range(0, len(s)):
            cutStr = s[start:start+cut]
            
            if isPalindrome(cutStr) == True:
                return len(cutStr)
            
            if start+cut >= len(s):
                break

# 시간초과 -> 길이가 작은 것 부터 찾으면 시간초과


# def solution(s):
#     answer = 0
#     d = {}
    
#     N = len(s)
    
#     max_len = 0
#     for k in range(N):
#         tmp = ''
#         for i in range(k, (N+k)//2 + 1):
#             tmp += s[i]
            
            
#             if isPalindrome(tmp):
#                 max_len = max(max_len, len(tmp))

#             cmp = ''
#             for j in range(i+1, N):
#                 if j >= N:
#                     break
#                 cmp += s[j]
                
#                 if tmp+cmp in d:
#                     continue

#                 if isPalindrome(tmp + cmp):
#                     if tmp+cmp not in d:
#                         d[tmp+cmp] = len(tmp+cmp)
                        
#                     max_len = max(max_len, len(tmp+cmp))
    
#     answer = max_len
#     return answer